lib/pbench/server/database/models/api_keys.py

```diff
@@ -0,0 +1,128 @@
+from typing import Optional
+
+from flask import current_app
+import jwt
+from sqlalchemy import Column, ForeignKey, String
+from sqlalchemy.orm import relationship
+
+from pbench.server.database.database import Database
+from pbench.server.database.models import decode_integrity_error, TZDateTime
+from pbench.server.database.models.users import User
+
+# Module private constants
+_TOKEN_ALG_INT = "HS256"
+
+
+class APIKeyError(Exception):
+    """A base class for errors reported by the APIKey class."""
+
+    def __init__(self, message):
+        self.message = message
+
+    def __str__(self):
+        return repr(self.message)
+
+
+class DuplicateApiKey(APIKeyError):
+    """Attempt to commit a duplicate unique value."""
+
+    def __init__(self, cause: str):
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Duplicate api_key: {self.cause}"
+
+
+class NullKey(APIKeyError):
+    """Attempt to commit an APIkey row with an empty required column."""
+
+    def __init__(self, cause: str):
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Missing required key: {self.cause}"
+
+
+class APIKey(Database.Base):
+    """Model for storing the API key associated with a user."""
+
+    __tablename__ = "api_keys"
+    api_key = Column(String(500), primary_key=True)
+    created = Column(TZDateTime, nullable=False, default=TZDateTime.current_time)
+    # ID of the owning user
+    user_id = Column(String, ForeignKey("users.id"), nullable=False)
+
+    # Indirect reference to the owning User record
+    user = relationship("User")
+
+    def __str__(self):
+        return f"key: {self.api_key}"
+
+    def add(self):
+        """Add an api_key object to the database."""
+        try:
+            Database.db_session.add(self)
+            Database.db_session.commit()
+        except Exception as e:
+            Database.db_session.rollback()
+            self.logger.error("Can't add key:{} to DB: {}", str(self), str(e))
+            raise decode_integrity_error(
+                e, on_duplicate=DuplicateApiKey, on_null=NullKey
+            )
```

Comment:
If we're going to raise a new exception here, then we should have a `from` clause on the `raise`.  Presumably, just `from e`, which, hopefully, will do the right thing in all the cases.

(I don't know whether `raise e from e` has the same semantics as a bare `raise` -- it would be handy if it did -- but, if it doesn't, it will at least have a better result than we would get omitting the `from`.)